src/main/java/org/opentripplanner/updater/vehicle_positions/VehiclePositionPatternMatcher.java

```diff
@@ -0,0 +1,205 @@
+            String feedId,
+            VehiclePosition vehiclePosition
+    ) {
+        if (!vehiclePosition.hasTrip()) {
+            LOG.warn("Realtime vehicle positions {} has no trip ID. Ignoring.", vehiclePosition);
+            return null;
+        }
+
+        String tripId = vehiclePosition.getTrip().getTripId();
+        Trip trip = getTripForId.apply(new FeedScopedId(feedId, tripId));
+        if (trip == null) {
+            LOG.warn(
+                    "Unable to find trip ID in feed '{}' for vehicle position with trip ID {}",
+                    feedId, tripId
+            );
+            return null;
+        }
+
+        var localDate =
+                Optional.of(vehiclePosition.getTrip().getStartDate())
+                        .map(Strings::emptyToNull)
+                        .flatMap(ServiceDate::parseStringToOptional);
+
+        if (localDate.isEmpty()) {
+            LOG.warn(
+                    "Trip with id {} doesn't contain a valid start_data which is required. Ignoring.",
+                    vehiclePosition.getTrip().getTripId()
+            );
+            return null;
+        }
+
+        TripPattern pattern = getPatternForTrip.apply(trip, localDate.get());
+        if (pattern == null) {
+            LOG.warn(
+                    "Unable to match OTP pattern ID for vehicle position with trip ID {}",
+                    tripId
+            );
+            return null;
+        }
+
+        // Add position to pattern
+        var newPosition = mapVehiclePosition(
+                vehiclePosition,
+                pattern.getStops(),
+                trip
+        );
+
+        return new T2<>(pattern, newPosition);
+    }
+
+    /**
+     * Converts GtfsRealtime vehicle position to the OTP RealtimeVehiclePosition which can be used
+     * by the API.
+     */
+    public static RealtimeVehiclePosition mapVehiclePosition(
+            VehiclePosition vehiclePosition,
+            List<StopLocation> stopsOnVehicleTrip,
+            Trip trip
+    ) {
+        var newPosition = RealtimeVehiclePosition.builder();
+
+        if (vehiclePosition.hasPosition()) {
+            var position = vehiclePosition.getPosition();
+            newPosition.setCoordinates(
+                    new WgsCoordinate(position.getLatitude(), position.getLongitude()));
+
+            if (position.hasSpeed()) {
+                newPosition.setSpeed(position.getSpeed());
+            }
+            if (position.hasBearing()) {
+                newPosition.setHeading(position.getBearing());
+            }
+        }
+
+        if (vehiclePosition.hasVehicle()) {
+            var vehicle = vehiclePosition.getVehicle();
+            newPosition.setVehicleId(vehicle.getId())
+                    .setLabel(Optional.ofNullable(vehicle.getLabel())
+                            .orElse(vehicle.getLicensePlate()));
+        }
+
+        if (vehiclePosition.hasCurrentStatus()) {
```

Comment:
After re-reading the spec I'm confused.

Why do you need both current_stop_sequence _and_ stop_id? What order of precedence do they have if they are both in a feed?

What do you think of the following, @optionsome?
- If stop_id is there, use that.
- Otherwise use stop_sequence.

Reply:
I don't know the history behind it but I also prefer to stop_id over the sequence. I don't think I have seen a feed that uses the stop_sequence.

Reply:
I've implemented the logic described above in ab6d3660efebfefe3317b50b9fe09f0a95f91231

The RealtimeVehiclePostionBuilder takes care that we only create a `stop` property when we either have a stop_id or current_stop_sequence.